### PR TITLE
MM-52845 : Migrate "components/admin_console/elasticsearch_settings.jsx" and tests to Typescript

### PR DESCRIPTION
**Pull request by karan2704**
#### Summary

Migrate components/admin_console/elasticsearch_settings.jsx to typescript
#### Ticket Link

Fixes https://github.com/mattermost/mattermost/issues/23415
Jira https://mattermost.atlassian.net/browse/MM-52845

#### Screenshots

N/A

#### Release Note

```release-note
N/A
```

### CODE REVIEW

**Comment by karan2704** [comment]
@M-ZubairAhmed I have addressed the requested changes in this PR: https://github.com/mattermost/mattermost/pull/23762/ which I closed due to a lot of conflicts.

image
Here Props['config'] is of type DeepPartial(AdminConfig) which shows an error for each config attribute saying "it might be undefined"

**Comment by karan2704** [comment]
image
Also, removing input makes it undefined, and although the type props in text_settings.tsx is optional and its type (InputTypes) can be assigned undefined it shows a compilation error saying `string|undefined cannot be assigned to type string` on this line
https://github.com/mattermost/mattermost/pull/23821/files#diff-03b7c0f5c200adcb1bb582065591572579850c61c596b12a90ca5f93cc834cc7L77

File: webapp/channels/src/components/admin_console/elasticsearch_settings.tsx [hunk 1]
```diff
@@ -36,7 +56,7 @@ export default class ElasticsearchSettings extends AdminSettings {
         return config;
     };
 
-    getStateFromConfig(config) {
+    getStateFromConfig(config: AdminConfig) {
```
Comment (M-ZubairAhmed) [inline webapp/channels/src/components/admin_console/elasticsearch_settings.tsx:59]: https://github.com/mattermost/mattermost/blob/8056be1d165c68e19dcda7429001055af9489d75/webapp/channels/src/components/admin_console/bleve_settings.tsx#L23-L25

Can you do similar to this. So it all ties back to Props

**Review by M-ZubairAhmed — CHANGES_REQUESTED** [review]
This is good, also you got conflicts again sorry :(

**Comment by karan2704** [comment]
@M-ZubairAhmed Can you point out the git practices used for mattermost... it seems to push a lot of unnecessary commits when I try to rebase.

**Comment by M-ZubairAhmed** [comment]
> @M-ZubairAhmed Can you point out the git practices used for mattermost... it seems to push a lot of unnecessary commits when I try to rebase.

Yes sorry I know Git can be very painful sometimes to deal with. I usually do `git pull origin master --no-rebase`. Another way to do if you don't want to deal with rebase and all is to run the `/update-branch` command.

**Comment by M-ZubairAhmed** [comment]
@karan2704 i have merged and resolved the conflicts for you. Can you please make another pass and remove if something unrelated creeped up in the PR

**Comment by karan2704** [comment]
Made the required changes ...please review

File: server/public/model/feature_flags.go
Comment (M-ZubairAhmed) [inline server/public/model/feature_flags.go:1]: can you revert this change?

File: webapp/channels/src/components/widgets/settings/text_setting.tsx [hunk 2]
```diff
@@ -5,7 +5,7 @@ import React from 'react';
 
 import Setting from './setting';
 
-export type InputTypes = 'input' | 'textarea' | 'number' | 'email' | 'tel' | 'url' | 'password'
+export type InputTypes = 'input' | 'textarea' | 'number' | 'email' | 'tel' | 'url' | 'password' | undefined
```
Comment (M-ZubairAhmed) [inline webapp/channels/src/components/widgets/settings/text_setting.tsx:8]: do you need to pass explicitly undefined here?

File: webapp/channels/src/components/widgets/settings/text_setting.tsx [hunk 3]
```diff
@@ -73,6 +73,21 @@ export default class TextSetting extends React.PureComponent<WidgetTextSettingPr
...
+                    id={this.props.id}
+                    className='form-control'
+                    type='input'
+                    placeholder={this.props.placeholder}
+                    value={this.props.value}
+                    maxLength={this.props.maxLength}
+                    onChange={this.handleChange}
+                    disabled={this.props.disabled}
+                />
+            );
```
Comment (M-ZubairAhmed) [inline webapp/channels/src/components/widgets/settings/text_setting.tsx:90]: so why did we add this here?

**Review by M-ZubairAhmed — CHANGES_REQUESTED** [review]

File: webapp/channels/src/components/widgets/settings/text_setting.tsx (hunk 2 above)
Comment (karan2704) [reply webapp/channels/src/components/widgets/settings/text_setting.tsx:8]: Since the type prop is optional if it isn't passed it will be undefined

File: webapp/channels/src/components/widgets/settings/text_setting.tsx (hunk 3 above)
Comment (karan2704) [reply webapp/channels/src/components/widgets/settings/text_setting.tsx:90]: in the else block the line `type = ['input', 'email', 'tel', 'number', 'url', 'password'].includes(type) ? type : 'input';` will show error if undefined is passed in includes function that is why I handled undefined case separately

Comment (M-ZubairAhmed) [reply webapp/channels/src/components/widgets/settings/text_setting.tsx:90]: if thats the case then lets pass correct `type` in TextSetting of `webapp/channels/src/components/admin_console/elasticsearch_settings.tsx`

**Review by M-ZubairAhmed — CHANGES_REQUESTED** [review]

File: webapp/channels/src/components/widgets/settings/text_setting.tsx (hunk 2 above)
Comment (M-ZubairAhmed) [reply webapp/channels/src/components/widgets/settings/text_setting.tsx:8]: we dont need to do this as per https://github.com/mattermost/mattermost/pull/23821#discussion_r1246011918

**Comment by karan2704** [comment]
As per this [comment](https://github.com/mattermost/mattermost/pull/23762#discussion_r1233336494) on my previously closed PR type is an optional prop and need not be passed every time but I have removed the undefined check and instead just added a no null assertion (!)

**Comment by M-ZubairAhmed** [comment]
> As per this [comment](https://github.com/mattermost/mattermost/pull/23762#discussion_r1233336494) on my previously closed PR type is an optional prop and need not be passed every time but I have removed the undefined check and instead just added a no null assertion (!)

Actually sorry this caused a confusion, I will send another PR to fix this. Other contributors have also faced something similar to this.

**Comment by karan2704** [comment]
Is this the same case for `isDisabled` prop in admin_console/admin_settings.tsx because that is causing the above tests to fail?

**Comment by M-ZubairAhmed** [comment]
> Is this the same case for `isDisabled` prop in admin_console/admin_settings.tsx because that is causing the above tests to fail?

Thats because this pr changes that props from it being not optional. Can we make it back to be optional.

File: webapp/channels/src/components/admin_console/push_settings.tsx [hunk 4]
```diff
@@ -54,7 +54,7 @@ export default class PushSettings extends AdminSettings<Props, State> {
         });
     };
 
-    handleDropdownChange = (id: string, value: string) => {
+    handleDropdownChange = (id: string, value: EmailSettings['PushNotificationServerType'] | EmailSettings['PushNotificationServerLocation']) => {
```
Comment (devinbinnie) [inline webapp/channels/src/components/admin_console/push_settings.tsx:57]: Seeing some type errors around this change:
```
Error: src/components/admin_console/push_settings.tsx(322,21): error TS2769: No overload matches this call.
  Overload 1 of 2, '(props: Props | Readonly<Props>): DropdownSetting', gave the following error.
    Type '(id: string, value: EmailSettings['PushNotificationServerType'] | EmailSettings['PushNotificationServerLocation']) => void' is not assignable to type '(id: string, value: string) => void'.
      Types of parameters 'value' and 'value' are incompatible.
        Type 'string' is not assignable to type '"off" | "custom" | "de" | "us" | "mhpns" | "mtpns"'.
  Overload 2 of 2, '(props: Props, context: any): DropdownSetting', gave the following error.
    Type '(id: string, value: EmailSettings['PushNotificationServerType'] | EmailSettings['PushNotificationServerLocation']) => void' is not assignable to type '(id: string, value: string) => void'.
Error: src/components/admin_console/push_settings.tsx(341,21): error TS2769: No overload matches this call.
  Overload 1 of 2, '(props: Props | Readonly<Props>): DropdownSetting', gave the following error.
    Type '(id: string, value: EmailSettings['PushNotificationServerType'] | EmailSettings['PushNotificationServerLocation']) => void' is not assignable to type '(id: string, value: string) => void'.
  Overload 2 of 2, '(props: Props, context: any): DropdownSetting', gave the following error.
    Type '(id: string, value: EmailSettings['PushNotificationServerType'] | EmailSettings['PushNotificationServerLocation']) => void' is not assignable to type '(id: string, value: string) => void'.
```

**Comment by karan2704** [comment]
Is there a way for me to run tests whenever committing new changes?

**Comment by M-ZubairAhmed** [comment]
> Is there a way for me to run tests whenever committing new changes?

Yes take a look at this https://developers.mattermost.com/contribute/more-info/webapp/developer-workflow/

**Comment by karan2704** [comment]
I am not able to see the details of the failed test above. It says page not found

**Comment by karan2704** [comment]
The failing `Enterprise CI/tests` goes to [this](https://github.com/mattermost/enterprise/actions/runs/5469013226) page when clicking on details. Do you know what could be going wrong here? @M-ZubairAhmed

**Comment by M-ZubairAhmed** [comment]
@karan2704 can you please update your fork from `mattermost-server` to `mattermost` https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/syncing-a-fork

That could be the reason for failing ee tests

File: webapp/channels/src/components/widgets/settings/text_setting.tsx
Comment (M-ZubairAhmed) [inline webapp/channels/src/components/widgets/settings/text_setting.tsx:1]: please revert the changes in this file, I have fixed this in another pr, you may merge the latest master

**Review by M-ZubairAhmed — CHANGES_REQUESTED** [review]

**Review by M-ZubairAhmed — APPROVED** [review]

File: webapp/channels/src/components/admin_console/elasticsearch_settings.tsx [hunk 5]
```diff
@@ -97,7 +121,7 @@ export default class ElasticsearchSettings extends AdminSettings {
         return this.state.canSave;
     };
 
-    doTestConfig = (success, error) => {
+    doTestConfig = (success: (data?: any) => void, error: (error: any) => void): void => {
```
Comment (devinbinnie) [inline webapp/channels/src/components/admin_console/elasticsearch_settings.tsx:124]: We can remove the parameter for `data` here since it isn't used.

Also `error` can be `string` instead of `any`.

**Review by devinbinnie — APPROVED** [review]
LGTM! Just left one non-blocking comment.
Thanks @karan2704

**Review by yasserfaraazkhan — APPROVED** [review]

**Comment by M-ZubairAhmed** [comment]
@karan2704 thank you so much again for contributing to mattermost with this PR. Your contribution is very valuable. 💙 🎊 We appreciate your help to help make better software.

**Comment by karan2704** [comment]
Thanks. I am happy to help.  😄
